hudi-spark-datasource/hudi-spark-common/src/test/java/org/apache/hudi/TestHoodieSchemaUtils.java

```diff
@@ -239,6 +240,51 @@ void testMissingColumn(boolean allowDroppedColumns) {
     }
   }
 
+  @Test
+  void testFieldReordering() {
+    // field order changes and incoming schema is missing an existing field
+    Schema start = createRecord("reorderFields",
+        createPrimitiveField("field1", Schema.Type.INT),
+        createPrimitiveField("field2", Schema.Type.INT),
+        createPrimitiveField("field3", Schema.Type.INT));
+    Schema end = createRecord("reorderFields",
+        createPrimitiveField("field3", Schema.Type.INT),
+        createPrimitiveField("field1", Schema.Type.INT));
+    assertEquals(start, deduceWriterSchema(end, start, true));
+
+    // nested field ordering changes and new field is added
+    start = createRecord("reorderNestedFields",
+        createPrimitiveField("field1", Schema.Type.INT),
+        createPrimitiveField("field2", Schema.Type.INT),
+        createArrayField("field3", createRecord("nestedRecord",
+            createPrimitiveField("nestedField1", Schema.Type.INT),
+            createPrimitiveField("nestedField2", Schema.Type.INT),
+            createPrimitiveField("nestedField3", Schema.Type.INT))),
+        createPrimitiveField("field4", Schema.Type.INT));
+    end = createRecord("reorderNestedFields",
+        createPrimitiveField("field1", Schema.Type.INT),
+        createPrimitiveField("field2", Schema.Type.INT),
+        createPrimitiveField("field5", Schema.Type.INT),
+        createArrayField("field3", createRecord("nestedRecord",
+            createPrimitiveField("nestedField2", Schema.Type.INT),
+            createPrimitiveField("nestedField1", Schema.Type.INT),
+            createPrimitiveField("nestedField3", Schema.Type.INT),
+            createPrimitiveField("nestedField4", Schema.Type.INT))),
+        createPrimitiveField("field4", Schema.Type.INT));
+
+    Schema expected = createRecord("reorderNestedFields",
+        createPrimitiveField("field1", Schema.Type.INT),
+        createPrimitiveField("field2", Schema.Type.INT),
+        createArrayField("field3", createRecord("reorderNestedFields.field3",
```

Comment:
@jonvex can you confirm this is the expected naming after reconcile is run?

Reply:
What should it look like?

Reply:
https://github.com/apache/hudi/blob/master/hudi-common/src/main/java/org/apache/hudi/internal/schema/convert/AvroInternalSchemaConverter.java#L341 - this logic was not created/updated by me. Do you want me to change it as part of this PR?

Reply:
ok, can you please change the nested record name to `reorderNestedFields.field3` in start and end? That way we isolate what we are testing